### PR TITLE
Feat response json

### PR DESCRIPTION
**Pull request by Arshia001**
This PR includes a fix to `#[js_class]`'s code generation for async functions with a `&Context` parameter, which is then used to implement `Response.json()`.

I'm not entirely sure about the error handling though: `return Err(ion::Error::new("Failed to deserialize JSON", ion::ErrorKind::Normal));`

### CODE REVIEW

**Comment by Redfire75369** [comment]
`JS_ParseJSON1` already sets the pending exception (as per how the engine deems), so you just need `Err(Error::none())`.

The changes all seem fine. Ironically though, I was working on removing async native functions/methods in favour of `promise_to_future` though, along with the changes to how classes work. It would probably be better if this landed after that change.

**Comment by Arshia001** [comment]
OK then, I'll wait until your merges are in.

File: runtime/src/globals/fetch/response/mod.rs [hunk 1]
```diff
@@ -176,5 +176,18 @@ pub mod class {
 			let bytes = self.read_to_bytes().await?;
 			String::from_utf8(bytes).map_err(|e| Error::new(&format!("Invalid UTF-8 sequence: {}", e), None))
 		}
+
+		pub async fn json(&mut self, cx: &Context<'_>) -> Result<*mut JSObject> {
```
Comment (Redfire75369) [inline runtime/src/globals/fetch/response/mod.rs:180]: The `'_` lifetime is unnecessary here.

File: runtime/src/globals/fetch/response/mod.rs [hunk 2]
```diff
@@ -176,5 +176,18 @@ pub mod class {
...
+			let text = self.text().await?;
+			let Some(str) = ion::String::new(cx, text.as_str()) else {
+				return Err(ion::Error::new("Failed to allocate string", ion::ErrorKind::Normal));
+			};
+			let mut result = ion::Value::undefined(cx);
+			if !unsafe { mozjs::jsapi::JS_ParseJSON1(cx.as_ptr(), str.handle().into(), result.handle_mut().into()) } {
+				return Err(ion::Error::new("Failed to deserialize JSON", ion::ErrorKind::Normal));
```
Comment (Redfire75369) [inline runtime/src/globals/fetch/response/mod.rs:187]: Use Err(Error::none()) here so the exception from the engine bubbles.

File: runtime/src/globals/fetch/response/mod.rs [hunk 3]
```diff
@@ -176,5 +176,18 @@ pub mod class {
 			let bytes = self.read_to_bytes().await?;
 			String::from_utf8(bytes).map_err(|e| Error::new(&format!("Invalid UTF-8 sequence: {}", e), None))
 		}
+
+		pub async fn json(&mut self, cx: &Context<'_>) -> Result<*mut JSObject> {
+			let text = self.text().await?;
+			let Some(str) = ion::String::new(cx, text.as_str()) else {
+				return Err(ion::Error::new("Failed to allocate string", ion::ErrorKind::Normal));
```
Comment (Redfire75369) [inline runtime/src/globals/fetch/response/mod.rs:183]: Using `.unwrap()` here is fine. If the engine runs out of memory, that's an uncatchable exception anyways.

**Review by redfire75369 — CHANGES_REQUESTED** [review]
Changes are mostly fine, just need to be adapted to the overhaul of classes.

**Pull request closed without merging**